File: Sources/ClaudeUsageWidgetCore/Usage/UsageMath.swift
import Foundation

/// Pure arithmetic behind the dials. No I/O, no clock reads — `now` is always
/// passed in so every case is reproducible in a test.
public enum UsageMath {
    /// Where the hand points: the wall-clock position of the reset time on a
    /// 12-hour dial, as 0...1 of a full revolution from twelve o'clock.
    /// A reset at 20:00 local time puts the hand where a watch's hour hand
    /// would sit at 8 o'clock.
    ///
    /// Nil when there is no reset time or it has already passed — a stale
    /// snapshot hides the hand rather than pointing at a wrong time.
    public static func clockFraction(resetsAt: Date?, now: Date, calendar: Calendar = .current) -> Double? {
        guard let resetsAt, resetsAt > now else { return nil }
        let parts = calendar.dateComponents([.hour, .minute, .second], from: resetsAt)
        let seconds = Double((parts.hour ?? 0) % 12) * 3600
            + Double(parts.minute ?? 0) * 60
            + Double(parts.second ?? 0)
        return seconds / (12 * 3600)
    }

    /// Time left in the window: "45s", "10m", "1h 0m", "1d 1h".
    /// Nil when there is no reset time or it has already passed.
    public static func remainingText(resetsAt: Date?, now: Date) -> String? {
        guard let resetsAt else { return nil }
        let seconds = Int(resetsAt.timeIntervalSince(now).rounded(.down))
        guard seconds > 0 else { return nil }

        let days = seconds / 86_400
        let hours = (seconds % 86_400) / 3600
        let minutes = (seconds % 3600) / 60

        if days > 0 { return "\(days)d \(hours)h" }
        if hours > 0 { return "\(hours)h \(minutes)m" }
        if minutes > 0 { return "\(minutes)m" }
        return "\(seconds)s"
    }

    /// The server reports utilization on a 0...100 scale; everything inside the
    /// widget works in 0...1.
    public static func fraction(_ utilization: Double) -> Double {
        min(max(utilization / 100, 0), 1)
    }
}

File: src/ClaudeUsageWidget.Core/Usage/UsageMath.cs
namespace ClaudeUsageWidget.Core;

/// Pure arithmetic behind the dials. No I/O, no clock reads — `now` is
/// always passed in from outside, so every case is reproducible in a test.
public static class UsageMath
{
    /// Time remaining until the window resets: "45s", "10m", "1h 0m", "1d 1h".
    /// Null if there's no reset time or it has already passed.
    public static string? RemainingText(DateTimeOffset? resetsAt, DateTimeOffset now)
    {
        if (resetsAt is not { } resets) return null;
        var seconds = (int)Math.Floor((resets - now).TotalSeconds);
        if (seconds <= 0) return null;

        var days = seconds / 86_400;
        var hours = (seconds % 86_400) / 3600;
        var minutes = (seconds % 3600) / 60;

        if (days > 0) return $"{days}d {hours}h";
        if (hours > 0) return $"{hours}h {minutes}m";
        if (minutes > 0) return $"{minutes}m";
        return $"{seconds}s";
    }

    /// The server reports utilization on a 0...100 scale; inside the widget
    /// everything works on a 0...1 scale.
    public static double Fraction(double utilization) =>
        Math.Min(Math.Max(utilization / 100, 0), 1);

    /// A 0...1 fraction as a string with a whole-number percent, e.g. "57%".
    ///
    /// Shifted by a tiny epsilon before rounding: a fraction obtained by
    /// dividing the exact server percentage by 100 can end up slightly below
    /// the .5 boundary due to binary representation — 0.575 * 100 equals
    /// 57.49999999999999 — and would round down against all expectations.
    /// The shift is far smaller than any real difference in the data.
    public static string PercentText(double fraction) =>
        $"{(int)Math.Round(fraction * 100 + 1e-9, MidpointRounding.AwayFromZero)}%";
}
